### PR TITLE
Support abandoned queued compilations

### PR DESCRIPTION
**Pull request by mattgodbolt**
After some time we know either the client or CloudFront will give up on pending compilations. As such, if we continue to process compilations after the client's timed out we're just clogging up the compilation queue with pointless work.

As such, this change now supports the notion of "stale" work which will be abandoned once it's made it to the front of the queue. Only compiles coming from the user will be abandoned, so discovery and health checks are unaffected.

Hopefully this will mitigate the number of nodes marked unhealthy due to being overloaded: work they were doing was "pointless" anyway, and them being killed by going unhealthy is equivalent to abandoning all the work in flight anyway, but this means there's a fighting chance the node will recover quickly enough to return a "healthy" status.

### CODE REVIEW

File: lib/compilation-queue.ts [hunk 1]
```diff
@@ -41,33 +41,59 @@ const queueCompleted = new PromClient.Counter({
...
         queueEnqueued.inc();
         return this._queue.add(() => {
+            const dequeuedAt = Date.now();
             queueDequeued.inc();
+            if (options && options.abandonIfStale && dequeuedAt > enqueuedAt + this._staleAfterMs) {
+                queueCompleted.inc();
+                queueStale.inc();
+                const queueTimeSecs = (dequeuedAt - enqueuedAt) / 1000;
+                const limitSecs = this._staleAfterMs / 1000;
+                throw new Error(
```
Comment (partouf) [inline lib/compilation-queue.ts:93]: how does the queue handler handle exceptions, console log and continue?

File: etc/config/compiler-explorer.defaults.properties [hunk 2]
```diff
@@ -42,6 +42,9 @@ remoteStorageServer=https://godbolt.org
 # it is set to 5 minutes so that it will never be triggered
 # under any normal circumstances
 compilationEnvTimeoutMs=300000
+# this timeout determines when a compilation in the queue is abandoned because it has been waiting too long and the
+# client will have given up on it by now
+compilationStaleAfterMs=100000
```
Comment (partouf) [inline etc/config/compiler-explorer.defaults.properties:47]: Is 100 seconds here arbitrary? Can we match it to compilationEnvTimeoutMs so the behaviour stays relatively the same?

File: lib/compilation-queue.ts (hunk 1 above)
Comment (mattgodbolt) [reply lib/compilation-queue.ts:93]: Yes.

File: etc/config/compiler-explorer.defaults.properties (hunk 2 above)
Comment (mattgodbolt) [reply etc/config/compiler-explorer.defaults.properties:47]: There's no change in behaviour with this setting: by being 100s in default it's essentially "forever". The real change is in the amazon properties.

Making it the same as compilationEnvTimeoutMs doesn't really make sense (to me) - the timeout there is mainly to catch a bug (as it says in the comment), not to actually time out the compiles (that's done with a much shorter timeout in the execution system).

**Comment by mattgodbolt** [comment]
Have done some testing in staging... it somewhat improves things, but with the health check being every 10s and "fail on 3 in a row" it really only gives us ~30s of leeway.

The error looks like:
```
 error: Error during compilation 2:  Compilation was in the queue too long (70.0s > 60.0s) {"stack":"Error: Compilation was in the queue too long (70.0s > 60.0s)\n    at file:///infra/.deploy/lib/compilation-queue.js:76:23\n    at file:///infra/.deploy/node_modules/p-queue/dist/index.js:111:37\n    at PQueue._PQueue_tryToStartAnother (file:///infra/.deploy/node_modules/p-queue/dist/index.js:285:13)\n    at PQueue._PQueue_next (file:///infra/.deploy/node_modules/p-queue/dist/index.js:237:85)\n    at file:///infra/.deploy/node_modules/p-queue/dist/index.js:131:88\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"} 
```

in papertrail.

I battered the one instance with timed-out requests and eventually got it to (briefly) be unhealthy still, but it did recover before it got nuked by EBS. So ... maybe an improvement?

We might consider making the abandon limit like 30s instead: if something's been queued that long it might be better to give up on it early as it'll possibly _still_ time out if it takes a while to compile. Thoughts?

**Comment by mattgodbolt** [comment]
Have drafted while we discuss more...I'd like to try _something_ to stem the endless cycling of machines, but maybe this isn't it.

**Comment by partouf** [comment]
> Have drafted while we discuss more...I'd like to try _something_ to stem the endless cycling of machines, but maybe this isn't it.

I'm not against the abandoning if the browser isn't likely listening anymore anyway, but this is entirely a separate thing from cycling. Sure it's way too busy on a node, but being busy is different from being unhealthy.

If you're busy you should retain everything that you've got and scale up.

I would suggest some kind of priority to healthcheck calls, it would be best to push those in front of the queue. Promise-queue doesn't support that, but it might be possible by forking it.

Or easier solution - just letting the healthcheck execute directly. Currently there's really no way that something runs for longer than desired, but of course when we add functionality we can make a mistake somewhere and introduce instances that lockup. But I don't think it's likely we'll make that mistake now that more and more code is async/await with lots of timeouts and more code is typescript. I don't think it's a gamble.

That + this PR would solve the issue of shutting down instances that are perfectly fine.

It does not solve the issue of scaling further up when queues get filled to the brim.

**Comment by partouf** [comment]
I just found out that p-queue also has an options parameter that you can give along with `.add()`

https://github.com/sindresorhus/p-queue#queue

So you can do `queue.add(healthCheckfunc, {priority: 99});` to put healthchecks in front of the queue.

**Comment by mattgodbolt** [comment]
> I'm not against the abandoning if the browser isn't likely listening anymore anyway, but this is entirely a separate thing from cycling. Sure it's way too busy on a node, but being busy is different from being unhealthy.

Not quite true: currently - on purpose - we put the health checks through the exact same queueing system to check there isn't a wedged compile in the queue. So...a few long compiles in the queue kill the queue.

>  I just found out that p-queue also has an options parameter that you can give along with .add()...

Right!...this might be a decent compromise too: it's still using the queue after all but with a "let it jump".

Abandoning the compilations is still useful though - in addition to a change like you describe. Otherwise we could end up with a "healthy" node that has 8 hours of queued compilations to churn through.

**Comment by partouf** [comment]
> Abandoning the compilations is still useful though - in addition to a change like you describe. Otherwise we could end up with a "healthy" node that has 8 hours of queued compilations to churn through.

Agreed

**Comment by mattgodbolt** [comment]
I'm going to try this out...and see what we end up with. Mea culpa

**Pull request merged**